File: packages/android/app/src/test/java/com/yepanywhere/mobile/experimental/ConversationClientTest.kt
package com.yepanywhere.mobile.experimental

import com.yepanywhere.mobile.connection.YaApiResponse
import java.net.URLDecoder
import kotlinx.coroutines.runBlocking
import org.json.JSONObject
import org.junit.Assert.*
import org.junit.Test

class ConversationClientTest {
    @Test fun missingAndDifferentRevisionsHaveDistinctFallbacks() {
        assertEquals(ConversationAvailability.UPDATE_REQUIRED, conversationAvailability(supported().put("experimentalSimpleClientApiRevision", JSONObject.NULL)))
        assertEquals(ConversationAvailability.REVISION_MISMATCH, conversationAvailability(supported().put("experimentalSimpleClientApiRevision", "future")))
    }

    private fun supported() = JSONObject("""{"capabilityBits":[[2,32]],"experimentalSimpleClientApiRevision":"simple-client-spike-1"}""")
    private fun frame(id: String, sequence: Int = 0) = JSONObject()
        .put("apiRevision", CONVERSATION_API_REVISION).put("subscriptionId", id).put("sequence", sequence)
        .put("view", JSONObject("""{"kind":"error","code":"unavailable","message":"Test source"}""")).toString()

    @Test fun missingCapabilityAndRevisionMismatchMakeZeroExperimentalCalls() = runBlocking {
        for (version in listOf(JSONObject("""{"current":"0.8.0"}"""), JSONObject("""{"current":"0.8.1"}"""), supported().put("experimentalSimpleClientApiRevision", JSONObject.NULL), supported().put("experimentalSimpleClientApiRevision", "future"))) {
            val paths = mutableListOf<String>()
            val client = ConversationClient("old") { path -> paths.add(path); YaApiResponse(200, emptyMap(), version) }
            assertTrue(client.read(ConversationQuery("s", 20, null)) is ConversationReadResult.Unavailable)
            assertEquals(listOf("/api/version"), paths)
        }
    }

    @Test fun healthyPeerConsumesTheTypedSnapshot() = runBlocking {
        val paths = mutableListOf<String>()
        val client = ConversationClient("peer") { path ->
            paths.add(path)
            val body = if (path == "/api/version") supported() else {
                val id = path.substringAfter("?").split("&").first { it.startsWith("subscriptionId=") }.substringAfter("=")
                JSONObject(frame(URLDecoder.decode(id, "UTF-8")))
            }
            YaApiResponse(200, emptyMap(), body)
        }
        val result = client.read(ConversationQuery("s", 20, null)) as ConversationReadResult.Snapshot
        assertEquals("peer", result.sourceId)
        assertEquals(0, result.snapshot.sequence)
        assertEquals(2, paths.size)
    }

    @Test fun bindingRejectsCrossSourceStaleDuplicateAndClosedFrames() {
        val binding = ConversationBinding("a", "new", "s")
        assertNull(binding.accept("b", frame("new")))
        assertNull(binding.accept("a", frame("old")))
        assertNotNull(binding.accept("a", frame("new")))
        assertNull(binding.accept("a", frame("new")))
        assertNotNull(binding.accept("a", frame("new", 2)))
        assertNull(binding.accept("a", frame("new", 1)))
        binding.close()
        assertNull(binding.accept("a", frame("new", 3)))
    }
}

File: packages/shared/test/conversation-client.test.ts
import { describe, expect, it, vi } from "vitest";
import {
  ConversationBinding,
  conversationAvailability,
  readConversation,
} from "../src/experimental/conversation-client.js";
import { CONVERSATION_API_REVISION } from "../src/experimental/conversation-protocol.js";
const supported = {
  capabilityEncoding: 1,
  capabilityBits: [[2, 32]] as const,
  experimentalSimpleClientApiRevision: CONVERSATION_API_REVISION,
};
const query = { sessionId: "s", maxMessages: 20, anchorMessageId: null };
const frame = (subscriptionId: string, sequence = 0) =>
  JSON.stringify({
    apiRevision: CONVERSATION_API_REVISION,
    subscriptionId,
    sequence,
    view: { kind: "error", code: "unavailable", message: "Test source" },
  });
describe("source-bound experimental Conversation client", () => {
  it("distinguishes missing revision metadata from a different preview revision", () => {
    expect(
      conversationAvailability({
        ...supported,
        experimentalSimpleClientApiRevision: undefined,
      }),
    ).toBe("update-required");
    expect(
      conversationAvailability({
        ...supported,
        experimentalSimpleClientApiRevision: "future",
      }),
    ).toBe("revision-mismatch");
  });

  it("makes zero calls for supported stable releases and revision mismatches while a peer succeeds", async () => {
    for (const version of [
      { current: "0.8.0" },
      { current: "0.8.1" },
      { ...supported, experimentalSimpleClientApiRevision: undefined },
      { ...supported, experimentalSimpleClientApiRevision: "future" },
    ]) {
      const request = vi.fn(async () => frame("binding"));
      const result = await readConversation({
        sourceId: "old",
        subscriptionId: "binding",
        query,
        version,
        request,
      });
      expect("unavailable" in result).toBe(true);
      expect(request).not.toHaveBeenCalled();
    }
    const request = vi.fn(async () => frame("binding"));
    expect(
      await readConversation({
        sourceId: "peer",
        subscriptionId: "binding",
        query,
        version: supported,
        request,
      }),
    ).toMatchObject({ sourceId: "peer", snapshot: { sequence: 0 } });
    expect(request).toHaveBeenCalledTimes(1);
  });
  it("rejects stale, duplicate, closed and cross-source frames even when session IDs collide", () => {
    const binding = new ConversationBinding("a", "new", "s");
    expect(binding.accept("b", frame("new"))).toBeNull();
    expect(binding.accept("a", frame("old"))).toBeNull();
    expect(binding.accept("a", frame("new"))).not.toBeNull();
    expect(binding.accept("a", frame("new"))).toBeNull();
    expect(binding.accept("a", frame("new", 2))).not.toBeNull();
    expect(binding.accept("a", frame("new", 1))).toBeNull();
    binding.close();
    expect(binding.accept("a", frame("new", 3))).toBeNull();
  });
});
